src/player.cpp

```diff
@@ -9039,82 +9039,59 @@ void player::use_fire(const int quantity)
+            auto bio = std::min( long( power_level ) , qty / 10 + ( qty % 10 != 0 ) );
+            charge_power( -bio );
+            qty -= std::min( qty, bio * 10 );
         }
-        return ret;
-    }
-    if (weapon.use_charges(it, quantity, ret)) {
-        remove_weapon();
-    }
-    for( auto a = worn.begin(); a != worn.end() && quantity > 0; ) {
-        if( a->use_charges( it, quantity, ret ) ) {
-            a = worn.erase( a );
-        } else {
-            ++a;
+
+        auto adv = charges_of( "adv_UPS_off", ceil( qty * 0.6 ) );
+        if( adv > 0 ) {
+            auto found = use_charges( "adv_UPS_off", adv );
+            res.splice( res.end(), found );
+            qty -= std::min( qty, long( adv / 0.6 ) );
+        }
+
+        auto ups = charges_of( "UPS_off", qty );
+        if( ups > 0 ) {
+            auto found = use_charges( "UPS_off", ups );
+            res.splice( res.end(), found );
+            qty -= std::min( qty, ups );
         }
     }
-    if (quantity <= 0) {
-        return ret;
-    }
-    std::list<item> tmp = inv.use_charges(it, quantity);
-    ret.splice(ret.end(), tmp);
-    if (quantity <= 0) {
-        return ret;
-    }
-    // Threat requests for UPS charges as request for adv. UPS charges
-    // and as request for bionic UPS charges, both with their own modificators
-    // If we reach this, the regular UPS could not provide all the requested
-    // charges, so we *have* to remove as many as charges as we can (but not
-    // more than requested) to not let any charges un-consumed.
-    if ( it == "UPS_off" ) {
-        // Request for 8 UPS charges:
-        // 8 UPS = 8 * 6 / 10 == 48/10 == 4.8 adv. UPS
-        // consume 5 adv. UPS charges, see player::charges_of, if the adv. UPS
-        // had only 4 charges, it would report as floor(4/0.6)==6 normal UPS
-        // charges
-        long quantity_adv = ceil(quantity * 0.6);
-        long avail_adv = charges_of("adv_UPS_off");
-        long adv_charges_to_use = std::min(avail_adv, quantity_adv);
-        if (adv_charges_to_use > 0) {
-            std::list<item> tmp = use_charges("adv_UPS_off", adv_charges_to_use);
-            ret.splice(ret.end(), tmp);
-            quantity -= static_cast<long>(adv_charges_to_use / 0.6);
-            if (quantity <= 0) {
-                return ret;
-            }
-        }
-    }
-    if ( power_level > 0 && it == "UPS_off" && has_active_bionic( "bio_ups" ) ) {
-        // Need always at least 1 power unit, to prevent exploits
-        // and make sure power_level does not get negative
-        long ch = std::max(1l, quantity / 10);
-        ch = std::min<long>(power_level, ch);
-        charge_power(-ch);
-        quantity -= ch * 10;
-        // TODO: add some(pseudo?) item to resulting list?
+
+    std::vector<item *> del;
+
+    visit_items( [this, &what, &qty, &res, &del]( item *e ) {
+        std::list<item> found;
+        if( e->use_charges( what, qty, found, &pos() ) ) {
```

Comment:
I was going to go with passing an output iterator but chose against that. Seems I missed considering the more obvious solution.